### PR TITLE
Added IHasUserContext abstraction

### PR DESCRIPTION
**Pull request by J-kit**
Makes it easier to write custom extension methods for the usercontext

### CODE REVIEW

File: src/GraphQL/Abstractions/IHasUserContext.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace GraphQL.Abstractions
```
Comment (sungam3r) [inline src/GraphQL/Abstractions/IHasUserContext.cs]: Do we need new namespace? Better move to `GraphQL.Execution`

File: src/GraphQL/Abstractions/IHasUserContext.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+{
+    public interface IHasUserContext
```
Comment (sungam3r) [inline src/GraphQL/Abstractions/IHasUserContext.cs]: There is already `IProvideMetadata` interface so this one can be also named `IProvideUserContext`

Comment (J-kit) [reply src/GraphQL/Abstractions/IHasUserContext.cs]: Hmm IHasX seems a bit more obvious, but if that's the convention im ok with that.

**Comment by J-kit** [comment]
> I will be glad to merge this after you fix the tests. `ApiApprovalTests.PublicApi.approved.txt` needs to be modified with new apis.

Ahh ok thanks, i didn't know this :P
Done

**Pull request merged**